### PR TITLE
NXP-19262 /   NXP-19264 /  NXP-26155: directory query improved

### PR DESCRIPTION
**Pull request by efge**
https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume/560/

### CODE REVIEW

File: nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-api/src/main/java/org/nuxeo/ecm/directory/Session.java [hunk 1]
```diff
@@ -198,8 +199,27 @@ DocumentModelList query(Map<String, Serializable> filter, Set<String> fulltext,
...
+     *
+     * @param queryBuilder the query to use, including limit, offset and ordering
+     * @param fetchReferences boolean stating if references have to be fetched
+     * @param countTotal if {@code true}, return a {@link DocumentModelList} that includes a total size of the
+     *            underlying list (the size if there was no limit or offset), otherwise does a best effort but may
+     *            return {@code -2} when unknown
+     * @return the list of documents
+     * @since 10.3
+     */
+    DocumentModelList query(QueryBuilder queryBuilder, boolean fetchReferences, boolean countTotal);
```
Comment (funshodavid) [inline nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-api/src/main/java/org/nuxeo/ecm/directory/Session.java]: Should we deprecate previous `query` methods in favor of this one ?

File: nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-ldap/src/main/java/org/nuxeo/ecm/directory/ldap/LDAPDirectory.java [hunk 2]
```diff
@@ -123,9 +126,11 @@ protected void initLDAPConfig() {
         // register the references
         addReferences(ldapDirectoryDesc.getLdapReferences());
 
+
```
Comment (funshodavid) [inline nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-ldap/src/main/java/org/nuxeo/ecm/directory/ldap/LDAPDirectory.java]: empty line not needed

File: nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-sql/src/main/java/org/nuxeo/ecm/directory/sql/SQLSession.java [hunk 3]
```diff
@@ -646,6 +673,245 @@ public DocumentModelList query(Map<String, Serializable> filter, Set<String> ful
...
+        }
+        if (FieldDetector.hasField(queryBuilder.predicate(), getPasswordField())) {
+            throw new DirectoryException("Cannot filter on password");
+        }
+        if (isMultiTenant()) {
+            // filter entries on the tenantId field also
+            String tenantId = getCurrentTenantId();
+            if (!StringUtils.isEmpty(tenantId)) {
+                // TODO don't modify passed-in value
+                queryBuilder.predicate(Predicates.eq(TENANT_ID_FIELD, tenantId));
```
Comment (kevinleturc) [inline nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-sql/src/main/java/org/nuxeo/ecm/directory/sql/SQLSession.java]: Forgotten TODO ? Can be tackled by `QueryBuilder` copy constructor right ?

File: nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-api/src/main/java/org/nuxeo/ecm/directory/BaseSession.java [hunk 4]
```diff
@@ -488,10 +491,37 @@ protected boolean canDeleteMultiTenantEntry(String entryId) {
...
+    /**
+     * Applies offset and limit to a List.
+     *
+     * @param list the original list
+     * @param limit maximum number of results, ignored if less than 1
+     * @param offset number of rows skipped before starting, will be 0 if less than 0
+     * @return the result with applied limit and offset
+     * @since 10.3
+     */
+    public <T> List<T> applyQueryLimits(List<T> list, int limit, int offset) {
```
[the remark this answers is not in the document]
Comment (efge) [reply nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-api/src/main/java/org/nuxeo/ecm/directory/BaseSession.java]: Currently the only caller is `queryIds` which itself is only used by `MultiDirectoryExpressionEvaluator.evaluate` which just needs a set of ids and doesn't have a use for the total size. I'd rather not complexify the API if it's not needed for an internal method.

File: nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-api/src/main/java/org/nuxeo/ecm/directory/Session.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (efge) [reply nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-api/src/main/java/org/nuxeo/ecm/directory/Session.java]: That's a good idea yes.

File: nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-sql/src/main/java/org/nuxeo/ecm/directory/sql/SQLSession.java (hunk 3 above)
Comment (efge) [reply nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-sql/src/main/java/org/nuxeo/ecm/directory/sql/SQLSession.java]: Yes I didn't have the copy constructor when I wrote this comment. Will change.

**Comment by kevinleturc** [comment]
👍

**Pull request merged**